File: Swift/Hash/[백준 1351] 무한 수열.swift
//
//  [백준 1351] 무한 수열.swift
//  Algorithm
//
//  Created by 김민 on 6/17/24.
//
// 백준 1351 무한 수열 https://www.acmicpc.net/problem/1351
//

/*
[📌 전략]
 피보나치 수열의 원리와 같음. 단 0 ≤ N ≤ 10^12인 점

[✅ 풀이]
 hash를 통해 풀이하기 위해 dictionary 사용
 이미 Ai의 답을 구한 적이 있다면 그 답을 그대로 재활용하여 시간 단축하기
 1. infinity라는 Ak를 구하는 함수 선언
 2. 딕셔너리에 k를 키값으로 가지는 요소가 존재하면 이미 답을 구한 것이므로 재활용
 3. 값이 없다면 재귀함수를 통해 dic[k] 구하기
*/

let line = readLine()!.split(separator: " ").map { Int($0)! }
let (n, p, q) = (line[0], line[1], line[2])
var dic = [Int: Int]()
dic[0] = 1


func infinity(_ k: Int) -> Int {
    if dic.contains(where: { $0.key == k }) {
        return dic[k]!
    } else {
        dic[k, default: 0] = infinity(k/p) + infinity(k/q)
        return dic[k]!
    }
}

print(infinity(n))

File: Python/Hash/[백준 1351] 무한 수열.py
'''
[백준 1351] 무한 수열
https://www.acmicpc.net/problem/1351

[📌 전략]
- 0 ≤ N ≤ 10^12 -> 리스트 사용 불가, 딕셔너리로 접근

[✅ 풀이]
1. 무한 수열의 형태를 그려보면 트리 구조로 나타낼 수 있고 이를 재귀 함수로 표현
2. 현재 cur이 딕셔너리에 존재할 시 그대로 cur의 value 리턴
3. 존재하지 않을 시 
    딕셔너리의 cur 키에 infinity(math.floor(cur/p)) + infinity(math.floor(cur/q)) value 
    저장 및 cur의 value 리턴

[📝 기록]
LV: 골드5
'''

import sys
import math

input = sys.stdin.readline
n, p, q = map(int, input().split())

my_dict = dict()
my_dict[0] = 1

def infinity(cur):
    if cur in my_dict:
        return my_dict[cur] 
    else:
        my_dict[cur] = infinity(math.floor(cur/p)) + infinity(math.floor(cur/q))
        return my_dict[cur]

print(infinity(n))